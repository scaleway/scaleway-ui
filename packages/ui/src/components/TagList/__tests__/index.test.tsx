import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import type { ReactNode } from 'react'
import { TagList } from '..'
import {
  renderWithTheme,
  shouldMatchEmotionSnapshot,
} from '../../../../.jest/helpers'

jest.mock('reakit/Tooltip', () => ({
  Tooltip: ({ children }: { children: ReactNode }) => children,
  TooltipArrow: () => null,
  TooltipReference: ({ children }: { children: ReactNode }) => children,
  useTooltipState: () => ({}),
}))

describe('TagList', () => {
  test('renders correctly', () =>
    shouldMatchEmotionSnapshot(
      <div>
        <TagList popoverTitle="Additional" tags={['scaleway', 'cloud']} />
      </div>,
    ))

  test('renders correctly with not tags', () =>
    shouldMatchEmotionSnapshot(
      <div>
        <TagList popoverTitle="Additional" />
      </div>,
    ))

  test('renders correctly with custom threshold', () =>
    shouldMatchEmotionSnapshot(
      <div>
        <TagList
          popoverTitle="Additional"
          threshold={2}
          tags={['scaleway', 'cloud']}
        />
      </div>,
    ))

  test('renders correctly with custom threshold and extra tags', () =>
    shouldMatchEmotionSnapshot(
      <div>
        <TagList
          popoverTitle="Additional"
          threshold={2}
          tags={['scaleway', 'cloud', 'provider']}
        />
      </div>,
    ))

  test('renders correctly with custom threshold and extra tags and maxLength inferior to the combined size of tags', () =>
    shouldMatchEmotionSnapshot(
      <div>
        <TagList
          popoverTitle="Additional"
          maxLength={10}
          threshold={2}
          tags={['scaleway', 'cloud', 'provider']}
        />
      </div>,
    ))

  test('renders correctly with multiline', () =>
    shouldMatchEmotionSnapshot(
      <div>
        <TagList
          popoverTitle="Additional"
          multiline
          threshold={2}
          tags={['scaleway', 'cloud', 'provider']}
        />
      </div>,
    ))

  test('renders correctly with copiable', () =>
    shouldMatchEmotionSnapshot(
      <div>
        <TagList
          popoverTitle="Additional"
          copiable
          threshold={2}
          tags={['scaleway', 'cloud']}
        />
      </div>,
    ))

  test('renders correctly with icons', () =>
    shouldMatchEmotionSnapshot(
      <div>
        <TagList
          popoverTitle="Additional"
          copiable
          threshold={4}
          tags={['scaleway', 'cloud']}
        />
      </div>,
    ))

  test('renders correctly when clicking on popover', async () => {
    renderWithTheme(
      <TagList
        popoverTitle="Additional"
        threshold={1}
        tags={[
          { label: 'smooth', icon: 'id' },
          'code',
          { label: 'hello', icon: 'lock' },
          { label: 'world', icon: 'plus' },
        ]}
      />,
    )

    expect(screen.queryByText('Additional')).not.toBeInTheDocument()

    const plus = screen.getByTestId('taglist-open')
    await userEvent.click(plus)

    expect(screen.getByText('Additional')).toBeInTheDocument()

    const closeButton = screen.getByLabelText('close')
    await userEvent.click(closeButton)

    await waitFor(() => {
      expect(screen.queryByText('Additional')).not.toBeInTheDocument()
    })
  })
})
