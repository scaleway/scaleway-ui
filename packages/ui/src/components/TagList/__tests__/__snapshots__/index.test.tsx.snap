// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagList renders correctly 1`] = `
<DocumentFragment>
  .cache-a1981g-StyledContainer {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.cache-1xwnq0e-StyledTagContainer {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #464959;
  gap: 8px;
}

.cache-1vdyh3h-StyledContainer {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  white-space: nowrap;
  border-radius: 4px;
  padding: 0 8px;
  gap: 8px;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  height: 24px;
  color: #464959;
  background: #ffffff;
  border: solid 1px #d9dadd;
}

.cache-t6vuyb-StyledTag {
  font-size: 12px;
  font-weight: 500;
  color: inherit;
  max-width: 232px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.cache-yb56yg-StyledChildrenContainer {
  display: inherit;
}

.cache-1wbx021-TagsWrapper {
  cursor: pointer;
  color: #521094;
  border: none;
  font-size: 14px;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  max-width: 350px;
  overflow: hidden;
  white-space: pre;
  text-overflow: ellipsis;
  background-color: transparent;
  padding-left: 8px;
  padding-right: 8px;
}

<div>
    <div
      class="cache-a1981g-StyledContainer eb6op482"
    >
      <div
        class="cache-1xwnq0e-StyledTagContainer eb6op480"
      >
        <span
          class="cache-1vdyh3h-StyledContainer el6733p2"
        >
          <span
            aria-disabled="false"
            class="cache-t6vuyb-StyledTag el6733p1"
          >
            scaleway
          </span>
        </span>
      </div>
      <div
        aria-describedby=":r0:"
        class="cache-yb56yg-StyledChildrenContainer e1h4zly10"
        tabindex="0"
      >
        <span
          class="cache-1wbx021-TagsWrapper eb6op481"
          data-testid="taglist-open"
        >
          +1
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TagList renders correctly with copiable 1`] = `
<DocumentFragment>
  .cache-a1981g-StyledContainer {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.cache-1xwnq0e-StyledTagContainer {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #464959;
  gap: 8px;
}

.cache-yb56yg-StyledChildrenContainer {
  display: inherit;
}

.cache-1vij52y-Container-StyledContainer {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  white-space: nowrap;
  border-radius: 4px;
  padding: 0 8px;
  gap: 8px;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  height: 24px;
  color: #464959;
  background: #ffffff;
  border: solid 1px #d9dadd;
}

.cache-1vij52y-Container-StyledContainer:hover,
.cache-1vij52y-Container-StyledContainer:active {
  cursor: pointer;
  background: #f3f3f4;
  border-color: #92959d;
}

.cache-1vij52y-Container-StyledContainer:active {
  box-shadow: 0px 0px 0px 3px #151A2D5C;
}

.cache-t6vuyb-StyledTag {
  font-size: 12px;
  font-weight: 500;
  color: inherit;
  max-width: 232px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

<div>
    <div
      class="cache-a1981g-StyledContainer eb6op482"
    >
      <div
        class="cache-1xwnq0e-StyledTagContainer eb6op480"
      >
        <div
          aria-describedby=":r4:"
          class="cache-yb56yg-StyledChildrenContainer e1h4zly10"
          tabindex="0"
        >
          <button
            class="cache-1vij52y-Container-StyledContainer el6733p3"
          >
            <span
              aria-disabled="false"
              class="cache-t6vuyb-StyledTag el6733p1"
            >
              scaleway
            </span>
          </button>
        </div>
        <div
          aria-describedby=":r5:"
          class="cache-yb56yg-StyledChildrenContainer e1h4zly10"
          tabindex="0"
        >
          <button
            class="cache-1vij52y-Container-StyledContainer el6733p3"
          >
            <span
              aria-disabled="false"
              class="cache-t6vuyb-StyledTag el6733p1"
            >
              cloud
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TagList renders correctly with custom threshold 1`] = `
<DocumentFragment>
  .cache-a1981g-StyledContainer {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.cache-1xwnq0e-StyledTagContainer {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #464959;
  gap: 8px;
}

.cache-1vdyh3h-StyledContainer {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  white-space: nowrap;
  border-radius: 4px;
  padding: 0 8px;
  gap: 8px;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  height: 24px;
  color: #464959;
  background: #ffffff;
  border: solid 1px #d9dadd;
}

.cache-t6vuyb-StyledTag {
  font-size: 12px;
  font-weight: 500;
  color: inherit;
  max-width: 232px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

<div>
    <div
      class="cache-a1981g-StyledContainer eb6op482"
    >
      <div
        class="cache-1xwnq0e-StyledTagContainer eb6op480"
      >
        <span
          class="cache-1vdyh3h-StyledContainer el6733p2"
        >
          <span
            aria-disabled="false"
            class="cache-t6vuyb-StyledTag el6733p1"
          >
            scaleway
          </span>
        </span>
        <span
          class="cache-1vdyh3h-StyledContainer el6733p2"
        >
          <span
            aria-disabled="false"
            class="cache-t6vuyb-StyledTag el6733p1"
          >
            cloud
          </span>
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TagList renders correctly with custom threshold and extra tags 1`] = `
<DocumentFragment>
  .cache-a1981g-StyledContainer {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.cache-1xwnq0e-StyledTagContainer {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #464959;
  gap: 8px;
}

.cache-1vdyh3h-StyledContainer {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  white-space: nowrap;
  border-radius: 4px;
  padding: 0 8px;
  gap: 8px;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  height: 24px;
  color: #464959;
  background: #ffffff;
  border: solid 1px #d9dadd;
}

.cache-t6vuyb-StyledTag {
  font-size: 12px;
  font-weight: 500;
  color: inherit;
  max-width: 232px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.cache-yb56yg-StyledChildrenContainer {
  display: inherit;
}

.cache-1wbx021-TagsWrapper {
  cursor: pointer;
  color: #521094;
  border: none;
  font-size: 14px;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  max-width: 350px;
  overflow: hidden;
  white-space: pre;
  text-overflow: ellipsis;
  background-color: transparent;
  padding-left: 8px;
  padding-right: 8px;
}

<div>
    <div
      class="cache-a1981g-StyledContainer eb6op482"
    >
      <div
        class="cache-1xwnq0e-StyledTagContainer eb6op480"
      >
        <span
          class="cache-1vdyh3h-StyledContainer el6733p2"
        >
          <span
            aria-disabled="false"
            class="cache-t6vuyb-StyledTag el6733p1"
          >
            scaleway
          </span>
        </span>
        <span
          class="cache-1vdyh3h-StyledContainer el6733p2"
        >
          <span
            aria-disabled="false"
            class="cache-t6vuyb-StyledTag el6733p1"
          >
            cloud
          </span>
        </span>
      </div>
      <div
        aria-describedby=":r1:"
        class="cache-yb56yg-StyledChildrenContainer e1h4zly10"
        tabindex="0"
      >
        <span
          class="cache-1wbx021-TagsWrapper eb6op481"
          data-testid="taglist-open"
        >
          +1
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TagList renders correctly with custom threshold and extra tags and maxLength inferior to the combined size of tags 1`] = `
<DocumentFragment>
  .cache-a1981g-StyledContainer {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.cache-1xwnq0e-StyledTagContainer {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #464959;
  gap: 8px;
}

.cache-1vdyh3h-StyledContainer {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  white-space: nowrap;
  border-radius: 4px;
  padding: 0 8px;
  gap: 8px;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  height: 24px;
  color: #464959;
  background: #ffffff;
  border: solid 1px #d9dadd;
}

.cache-t6vuyb-StyledTag {
  font-size: 12px;
  font-weight: 500;
  color: inherit;
  max-width: 232px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.cache-yb56yg-StyledChildrenContainer {
  display: inherit;
}

.cache-1wbx021-TagsWrapper {
  cursor: pointer;
  color: #521094;
  border: none;
  font-size: 14px;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  max-width: 350px;
  overflow: hidden;
  white-space: pre;
  text-overflow: ellipsis;
  background-color: transparent;
  padding-left: 8px;
  padding-right: 8px;
}

<div>
    <div
      class="cache-a1981g-StyledContainer eb6op482"
    >
      <div
        class="cache-1xwnq0e-StyledTagContainer eb6op480"
      >
        <span
          class="cache-1vdyh3h-StyledContainer el6733p2"
        >
          <span
            aria-disabled="false"
            class="cache-t6vuyb-StyledTag el6733p1"
          >
            scaleway
          </span>
        </span>
      </div>
      <div
        aria-describedby=":r2:"
        class="cache-yb56yg-StyledChildrenContainer e1h4zly10"
        tabindex="0"
      >
        <span
          class="cache-1wbx021-TagsWrapper eb6op481"
          data-testid="taglist-open"
        >
          +2
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TagList renders correctly with multiline 1`] = `
<DocumentFragment>
  .cache-a1981g-StyledContainer {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.cache-1wt2n1m-StyledTagContainer {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #464959;
  gap: 8px;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.cache-1vdyh3h-StyledContainer {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  white-space: nowrap;
  border-radius: 4px;
  padding: 0 8px;
  gap: 8px;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  height: 24px;
  color: #464959;
  background: #ffffff;
  border: solid 1px #d9dadd;
}

.cache-t6vuyb-StyledTag {
  font-size: 12px;
  font-weight: 500;
  color: inherit;
  max-width: 232px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.cache-yb56yg-StyledChildrenContainer {
  display: inherit;
}

.cache-1wbx021-TagsWrapper {
  cursor: pointer;
  color: #521094;
  border: none;
  font-size: 14px;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  max-width: 350px;
  overflow: hidden;
  white-space: pre;
  text-overflow: ellipsis;
  background-color: transparent;
  padding-left: 8px;
  padding-right: 8px;
}

<div>
    <div
      class="cache-a1981g-StyledContainer eb6op482"
    >
      <div
        class="cache-1wt2n1m-StyledTagContainer eb6op480"
      >
        <span
          class="cache-1vdyh3h-StyledContainer el6733p2"
        >
          <span
            aria-disabled="false"
            class="cache-t6vuyb-StyledTag el6733p1"
          >
            scaleway
          </span>
        </span>
        <span
          class="cache-1vdyh3h-StyledContainer el6733p2"
        >
          <span
            aria-disabled="false"
            class="cache-t6vuyb-StyledTag el6733p1"
          >
            cloud
          </span>
        </span>
      </div>
      <div
        aria-describedby=":r3:"
        class="cache-yb56yg-StyledChildrenContainer e1h4zly10"
        tabindex="0"
      >
        <span
          class="cache-1wbx021-TagsWrapper eb6op481"
          data-testid="taglist-open"
        >
          +1
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TagList renders correctly with not tags 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;
